_episodes/05-preprocessor.md

```diff
@@ -18,391 +18,507 @@ keypoints:
+> >          short_name: tas
+> > -        preprocessor: global_anomalies
+> > +        preprocessor: anomalies_amsterdam
+> >      scripts:
+> >        warming_stripes_script:
+> >          script: ~/eucp-project/warming_stripes.py
+> > ```
+> >
+> {: .solution}
+{:.challenge}
 
-You should be able to see the variables grouped under different subdirectories
-under your output preproc directory. The different groupings can be accessed in
-your diagnostic by selecting the key name of the field
-`variable_group`  such as `TAS_CONTROL`, or `TAS_TARGET`.
+> ## Different periods
+>
+> Split the diagnostic in 2: the second one should use a different period.
+> You're free to choose the periods yourself. For example, 1 could be 'recent',
+> the other '20th_century'. For this, you'll have to add a new variable group.
+>
+> > ## Solution
+> >
+> > Here's a copy of the [recipe at this point](../files/recipe_warming_stripes_periods.yml)
+> > and this is the difference with the previous recipe:
+> >
+> > ```diff
+> > --- files/recipe_warming_stripes_local.yml
+> > +++ files/recipe_warming_stripes_periods.yml
+> > @@ -7,7 +7,7 @@
+> >      - righi_mattia
+> >
+> >  datasets:
+> > -  - {dataset: BCC-ESM1, project: CMIP6, mip: Amon, exp: historical, ensemble: r1i1p1f1, grid: gn, start_year: 1850, end_year: 2014}
+> > +  - {dataset: BCC-ESM1, project: CMIP6, mip: Amon, exp: historical, ensemble: r1i1p1f1, grid: gn}
+> >
+> >  preprocessors:
+> >    anomalies_amsterdam:
+> > @@ -28,9 +28,16 @@
+> >  diagnostics:
+> >    diagnostic_warming_stripes:
+> >      variables:
+> > -      temperature_anomalies_amsterdam:
+> > +      temperature_anomalies_recent:
+> >          short_name: tas
+> >          preprocessor: anomalies_amsterdam
+> > +        start_year: 1950
+> > +        end_year: 2014
+> > +      temperature_anomalies_20th_century:
+> > +        short_name: tas
+> > +        preprocessor: anomalies_amsterdam
+> > +        start_year: 1900
+> > +        end_year: 1999
+> >      scripts:
+> >        warming_stripes_script:
+> >          script: ~/eucp-project/warming_stripes.py
+> > ```
+> >
+> {: .solution}
+{:.challenge}
 
-> ## How to find what CMIP data is available?
+> ## Different preprocessors
+>
+> Now that you have different variable groups, we can also use different
+> preprocessors. Add a second preprocessor to add another location of your
+> choosing.
+>
+> Pro-tip: if you want to avoid repetition, you can use YAML anchors.
+>
+> > ## Solution
+> >
+> > Here's a copy of the [recipe at this
+> > point](../files/recipe_warming_stripes_multiple_locations.yml) and this is
+> > the difference with the previous recipe:
+> >
+> > ```diff
+> > --- files/recipe_warming_stripes_periods.yml
+> > +++ files/recipe_warming_stripes_multiple_locations.yml
+> > @@ -14,7 +14,7 @@
+> >      extract_point:
+> >        latitude: 52.379189
+> >        longitude: 4.899431
+> > -    anomalies:
+> > +    anomalies: &anomalies
+> >        period: month
```

Comment:
I get this, but I don't know why:
```
Traceback (most recent call last):
  File "/home/stef/r/esmvalcore/esmvalcore/_main.py", line 430, in run
    fire.Fire(ESMValTool())
  File "/home/stef/miniconda3/envs/esmvaltool/lib/python3.8/site-packages/fire/core.py", line 138, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/home/stef/miniconda3/envs/esmvaltool/lib/python3.8/site-packages/fire/core.py", line 463, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/home/stef/miniconda3/envs/esmvaltool/lib/python3.8/site-packages/fire/core.py", line 672, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/home/stef/r/esmvalcore/esmvalcore/_main.py", line 407, in run
    process_recipe(recipe_file=recipe, config_user=cfg)
  File "/home/stef/r/esmvalcore/esmvalcore/_main.py", line 94, in process_recipe
    recipe = read_recipe_file(recipe_file, config_user)
  File "/home/stef/r/esmvalcore/esmvalcore/_recipe.py", line 71, in read_recipe_file
    return Recipe(raw_recipe,
  File "/home/stef/r/esmvalcore/esmvalcore/_recipe.py", line 951, in __init__
    self.tasks = self.initialize_tasks() if initialize_tasks else None
  File "/home/stef/r/esmvalcore/esmvalcore/_recipe.py", line 1299, in initialize_tasks
    task = _get_preprocessor_task(
  File "/home/stef/r/esmvalcore/esmvalcore/_recipe.py", line 917, in _get_preprocessor_task
    task = _get_single_preprocessor_task(
  File "/home/stef/r/esmvalcore/esmvalcore/_recipe.py", line 776, in _get_single_preprocessor_task
    products = _get_preprocessor_products(
  File "/home/stef/r/esmvalcore/esmvalcore/_recipe.py", line 756, in _get_preprocessor_products
    product.check()
  File "/home/stef/r/esmvalcore/esmvalcore/preprocessor/__init__.py", line 291, in check
    check_preprocessor_settings(self.settings)
  File "/home/stef/r/esmvalcore/esmvalcore/preprocessor/__init__.py", line 156, in check_preprocessor_settings
    raise ValueError(
ValueError: Unknown preprocessor function 'period', choose from: download, fix_file, load, derive, fix_metadata, concatenate, cmor_check_metadata, extract_time, extract_season, extract_month, fix_data, cmor_check_data, extract_levels, weighting_landsea_fraction, mask_landsea, mask_glaciated, mask_landseaice, regrid, extract_point, mask_fillvalues, mask_above_threshold, mask_below_threshold, mask_inside_range, mask_outside_range, clip, extract_region, extract_shape, extract_volume, extract_trajectory, extract_transect, detrend, multi_model_statistics, extract_named_regions, depth_integration, area_statistics, volume_statistics, amplitude, zonal_statistics, meridional_statistics, daily_statistics, monthly_statistics, seasonal_statistics, annual_statistics, decadal_statistics, climate_statistics, anomalies, regrid_time, timeseries_filter, linear_trend, linear_trend_stderr, convert_units, save, cleanup
```

Reply:
This is my recipe, I tried copy-pasting from the solution, but I get the same result:

```yaml
preprocessors:
  anomalies_amsterdam:
    extract_point:
      latitude: 52.379189
      longitude: 4.899431
      scheme: linear
    anomalies: &anomalies
      period: month
      reference:
        start_year: 1981
        start_month: 1
        start_day: 1
        end_year: 2010
        end_month: 12
        end_day: 31
      standardize: false
  anomalies_london:
    extract_point:
      latitude: 51.5074
      longitude: 0.1278
    <<: *anomalies

diagnostics:
  diagnostic_warming_stripes:
    description: visualize global temperature anomalies as warming stripes
    variables:
      temperature_anomalies_recent_amsterdam:
        short_name: tas
        preprocessor: anomalies_amsterdam
        start_year: 1950
        end_year: 2014
      temperature_anomalies_20th_century_london:
        short_name: tas
        preprocessor: anomalies_london
        start_year: 1900
        end_year: 1999
        additional_datasets:
          - {dataset: CanESM2, project: CMIP5, mip: Amon, exp: historical, ensemble: r1i1p1}
    scripts:
      warming_stripes_script:
        script:  /home/stef/s/stripes.py
        colormap: 'bwr'
```